upgrade/upgrade-scripts/rootfs-cleanup

```diff
@@ -103,67 +103,44 @@ def main() -> None:
     # currently running version, and the version used prior to the
     # current one.
     #
-    # By sorting the list of all filesystems and snapshots on the sytem,
+    #
+    # By sorting the list of all filesystems and snapshots on the system,
     # we can easily determine the highest two versions on the system,
     # and skip those versions (to maintain the policy described above);
     # this is based on the assumption that the versions of an appliance
     # will always increase (or remain constant, in the case of rollback)
     # after an upgrade (i.e. we never upgrade to a lower version).
     #
-    for rootfs in sorted(filesystems + snapshots, key=rootfscmp)[:-2]:
-        #
-        # In the event of a rollback, we want to be careful to not
-        # delete the currently running version, as well as any versions
-        # that may have failed to upgrade resulting in the rollback
-        # event to begin with.
-        #
-        # Additionally, we want to keep all rootfs snapshots that map to
-        # hotfixes for the currently running version. This way, we don't
-        # inadvertently remove a rootfs snapshot that we may need for
-        # root-causing a given problem; at least, not until we upgraded
-        # off of the problematic delphix version.
-        #
-        # Thus, to account for these cases, we skip any version that's
-        # equal to or greater than what's currently running.
-        #
-        #
+    # In the event of a rollback, we want to be careful to not
+    # delete the currently running version, as well as any versions
+    # that may have failed to upgrade resulting in the rollback
+    # event to begin with.
+    #
+    # Additionally, we want to keep all rootfs snapshots that map to
+    # hotfixes for the currently running version. This way, we don't
+    # inadvertently remove a rootfs snapshot that we may need for
+    # root-causing a given problem; at least, not until we upgraded
+    # off of the problematic delphix version.
+    #
+    # Thus, to account for these cases, we skip any version that's
+    # equal to or greater than what's currently running.
+    #
+    # Snapshots must be deleted before filesystems because the following code
+    # passes the `-r` option to the `zfs destroy` command to destroy a
+    # filesystem.
+    #
+    for rootfs in sorted(snapshots, key=rootfscmp)[:-2]:
```

Comment:
I'm not necessarily opposed to this approach (i.e. doing snapshots and filesystems seperately).. but this way, it looks like we'll (potentially) keep 2 snapshots and 2 filesystems.. which might equate to 4 versions, instead of the desired 2, right? which might result in more rootfs space being consumed by an additional 2 versions..?

Reply:
each snapshot or filesystem, represents one version.. so that's why we took the most recent 2, from the combination of both.. it doesn't matter if it's a filesystem or a snapshot.

for example, for the common case (e.g. no rollback, no lingering upgrade containers), we'll have a single filesystem with multiple snapshots on it.. each snapshot represents an old version.. so in this case, we'll generally keep the currently running root filesystem (i.e. the current version), and one snapshot on that filesystem (i.e. the previous version).. all other snapshots will represent even older versions, so we delete those..

Reply:
with that said, the 1 filesystem and 1 snapshot is just "coincidence".. e.g. if we happened to use not-in-place upgrade, I think we could wind up with 2 filesystems.. we just haven't actually used that mode of upgrade..

by combining all snapshots and filesystems (with each one uniquely representing a delphix version), and then selecting the most recent 2 from that combined list, we get our desired affect.. of only keeping the most recent 2 versions.. regardless of if those versions are a snapshot or a filesystem..

I think that was the intention, at least..

Reply:
makes sense. I'll change the approach then